src/lib/rnp2.cpp

```diff
@@ -3226,180 +3294,775 @@ rnp_generate_key_json(rnp_ffi_t ffi, const char *json, char **results)
     return ret;
 }
 
+static rnp_result_t
+rnp_generate_key_ex(rnp_ffi_t         ffi,
+                    const char *      alg,
+                    const char *      salg,
+                    uint32_t          bits,
+                    uint32_t          sbits,
+                    const char *      curve,
+                    const char *      scurve,
+                    const char *      userid,
+                    rnp_key_handle_t *key)
+{
+    rnp_op_generate_t op = NULL;
+    rnp_op_generate_t subop = NULL;
+    rnp_key_handle_t  primary = NULL;
+    rnp_result_t      ret = RNP_ERROR_KEY_GENERATION;
+
+    /* generate primary key */
+    if ((ret = rnp_op_generate_create(&op, ffi, alg))) {
+        return ret;
+    }
+
+    if (bits && (ret = rnp_op_generate_set_bits(op, bits))) {
+        goto done;
+    }
+    if (curve && (ret = rnp_op_generate_set_curve(op, curve))) {
+        goto done;
+    }
+    if ((ret = rnp_op_generate_set_userid(op, userid))) {
+        goto done;
+    }
+    if ((ret = rnp_op_generate_execute(op))) {
+        goto done;
+    }
+    if ((ret = rnp_op_generate_get_key(op, &primary))) {
+        goto done;
+    }
+    /* generate subkey if requested */
+    if (!salg) {
+        goto done;
+    }
+    if ((ret = rnp_op_generate_subkey_create(&subop, ffi, primary, salg))) {
+        goto done;
+    }
+    if (sbits && (ret = rnp_op_generate_set_bits(subop, sbits))) {
+        goto done;
+    }
+    if (scurve && (ret = rnp_op_generate_set_curve(subop, scurve))) {
+        goto done;
+    }
+    if ((ret = rnp_op_generate_execute(subop))) {
+        goto done;
+    }
+done:
+    if (!ret && key) {
+        *key = primary;
```

Comment:
* It looks like `key` is an optional arg (which is fine IMO) but will leak if NULL is provided and generation succeeds. I guess you would probably want to do `if (!key) { rnp_key_handle_destroy(primary); }`.
* Adding a line to a test for this would be good too, something like:
`assert_rnp_success(rnp_generate_key_rsa(ffi, 1024, 1024, "test_key_null", NULL));`
* Some clarification in the docs on who frees what would be ideal.

Reply:
Thanks for noticing this. Will fix. What do you think about making this function public? Make this any sense, or will give more questions from the users?

Reply:
@ni4 I could see it being useful as a public function. We can just clarify in the docs that it is an alternative to other methods of key generation.
(I'd prefer we make the param names more verbose/descriptive if public though)

Reply:
@dewyatt Yeah, good. What about adding password parameter as well, so key/subkey will go out encrypted after the single call?

Reply:
@ni4 That sounds like a good move, I guess `NULL` would be "no protection"?

Reply:
@dewyatt Yeah, exactly.

Reply:
@dewyatt I pushed some commits which address this thread and thread below with `rnp_op_generate_set_protection_password` instead of password provider.
Will address other comments a bit later.